### PR TITLE
Autobahn WebSocket server

### PR DESCRIPTION
**Pull request by mvollrath**
Instead of using obscure threading hackery to add flow control to Tornado, use a grown-up WebSocket server with its own flow control features.

### CODE REVIEW

**Comment by mvollrath** [comment]
CI pending https://github.com/ros/rosdistro/pull/21999

**Comment by uahic** [comment]
Since Autobahn seems also to have CBOR compression and all of that, do we need my pull request to add the support for compression in service calls then?

**Comment by mvollrath** [comment]
> Since Autobahn seems also to have CBOR compression and all of that, do we need my pull request to add the support for compression in service calls then?

This should not affect anything that happens in rosbridge_library, including protocol-level payload compression/encoding.

File: rosbridge_server/src/rosbridge_server/websocket_handler.py [hunk 1]
```diff
@@ -68,11 +66,37 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
-class RosbridgeWebSocket(WebSocketHandler):
+@implementer(interfaces.IPushProducer)
+class OutgoingValve:
+    """Allows the Autobahn transport to pause outgoing messages from rosbridge."""
```
Comment (viktorku) [inline rosbridge_server/src/rosbridge_server/websocket_handler.py]: Does this detect back-pressure and pauses publishing internally?

File: rosbridge_server/launch/rosbridge_websocket.launch [hunk 2]
```diff
@@ -1,6 +1,6 @@
 <launch>
   <arg name="port" default="9090" />
-  <arg name="address" default="" />
+  <arg name="address" default="0.0.0.0" />
```
Comment (mvollrath) [inline rosbridge_server/launch/rosbridge_websocket.launch]: A point of self-review: we need to be able to handle an empty "address" param for backwards compatibility.

File: rosbridge_server/src/rosbridge_server/websocket_handler.py [hunk 3]
```diff
@@ -68,11 +66,37 @@ def wrapper(*args, **kwargs):
...
+        reactor.callFromThread(self._proto.outgoing, message)
+
+    def pauseProducing(self):
+        self._valve.clear()
+
+    def resumeProducing(self):
+        self._valve.set()
+
+    def stopProducing(self):
+        rospy.loginfo('stop')
```
[the remark this answers is not in the document]
Comment (mvollrath) [reply rosbridge_server/src/rosbridge_server/websocket_handler.py]: Oops, this was left here accidentally.

**Comment by mvollrath** [comment]
CI now pending https://github.com/ros/rosdistro/pull/22032 because I forgot to add xenial.  Melodic tests are green.

I've done my own anecdotal testing on a couple of different use cases (point clouds and a very busy graph of small messages across many clients) which led to some fixes.  I may add some smoke tests for the server as a first step in this PR, mostly to make sure ssl works as configured.

File: rosbridge_server/src/rosbridge_server/websocket_handler.py (hunk 1 above)
Comment (mvollrath) [reply rosbridge_server/src/rosbridge_server/websocket_handler.py]: Yes, added some more description for this.

File: rosbridge_server/scripts/rosbridge_websocket.py [hunk 4]
```diff
@@ -238,20 +241,39 @@ def shutdown_hook():
...
     connected = False
     while not connected and not rospy.is_shutdown():
         try:
-            if certfile is not None and keyfile is not None:
-                application.listen(port, address, ssl_options={ "certfile": certfile, "keyfile": keyfile})
-            else:
-                application.listen(port, address)
-            rospy.loginfo("Rosbridge WebSocket server started on port %d", port)
+            listenWS(factory, context_factory)
+            rospy.loginfo('listening at {}'.format(uri))
```
[the remark this answers is not in the document]
Comment (mvollrath) [reply rosbridge_server/scripts/rosbridge_websocket.py]: This was done.

File: rosbridge_server/scripts/rosbridge_websocket.py [hunk 5]
```diff
@@ -52,18 +54,21 @@
 
 
 def shutdown_hook():
-    IOLoop.instance().stop()
+    try:
+        reactor.stop()
+    except ReactorNotRunning:
+        pass
```
[the remark this answers is not in the document]
Comment (mvollrath) [reply rosbridge_server/scripts/rosbridge_websocket.py]: Done.

File: rosbridge_server/package.xml [hunk 6]
```diff
@@ -15,9 +15,7 @@
   <maintainer email="jihoonlee.in@gmail.com">Jihoon Lee</maintainer>
 
   <buildtool_depend>catkin</buildtool_depend>
-  <exec_depend>python-backports.ssl-match-hostname</exec_depend>
-  <exec_depend>python-tornado</exec_depend>
-  <exec_depend>python-twisted-core</exec_depend>
```
[the remark this answers is not in the document]
Comment (mvollrath) [reply rosbridge_server/package.xml]: Good catch, fixed.

**Comment by jihoonl** [comment]
It seems like a huge ABI change which makes me concern about backward compatibility. Instead of replacing, how about making autobahn websocket server as addition to tornado? Then, I would add deprecation sign in tornado server that guides users to use autobahn instead.

**Comment by mvollrath** [comment]
> It seems like a huge ABI change which makes me concern about backward compatibility. Tornado has been too long to remove immediately.
> 
> Instead of replacing, how about making autobahn websocket server as addition to tornado? Then, I would add deprecation sign in tornado server that guides users to use autobahn instead.

I've made the Autobahn server the default and only script entrypoint.  The old Tornado handler is still there if anybody was using that interface.  The Autobahn server script supports the same features as the Tornado server (and a few more).

**Comment by mvollrath** [comment]
@dhananjaysathe thanks for your feedback, good to know this is already fixing things for people. :+1:

**Comment by mvollrath** [comment]
@viktorku I've made the smoke test run 100 messages of 10,000 byte strings in each direction, which always causes backpressure when I test it here (a protocol queue_length of 1 always drops some messages).

@jihoonl I've preserved the Tornado ABI for compatibility.

**Comment by jihoonl** [comment]
FYI @jspricke @mgrrx, A big change has been merged into develop branch. You may want to check if it breaks your usecase.

**Comment by jihoonl** [comment]
and thanks @mvollrath for nice updates.

**Comment by mgrrx** [comment]
@jihoonl thanks for the heads-up. We'll test it in the coming days.

**Comment by dirk-thomas** [comment]
I would be great of this patch would also be ported to the `ros2` branch so that it can benefit from this. I took a look at the diff but am not familiar enough with this to provide the PR myself.

**Comment by wolfv** [comment]
thanks, this is great. the tornado used by rosbridge was getting old and uncompatible with newer tornado versions as used in Jupyter, and websockets were silently failing :+1:

**Comment by moriarty** [comment]
> 

> > It seems like a huge ABI change which makes me concern about backward compatibility. Tornado has been too long to remove immediately.
> > Instead of replacing, how about making autobahn websocket server as addition to tornado? Then, I would add deprecation sign in tornado server that guides users to use autobahn instead.
> 
> I've made the Autobahn server the default and only script entrypoint. The old Tornado handler is still there if anybody was using that interface. The Autobahn server script supports the same features as the Tornado server (and a few more).

It would have been great if this breaking change had waited to go into Noetic, or defaulted the other way around to not introduce the breaking change which was warned about above.

This breaks compatibility with https://github.com/rctoris/jrosbridge
Admittedly it's breaking things which were under specified because javax.websockets isn't setting the scheme in the header Origin.

https://github.com/crossbario/autobahn-python/blob/13949d383af593fb4a78ed21079742c8234d67b1/autobahn/websocket/protocol.py#L96

Adding another proxy between the rosbridge and code that was previously working which adds the protocol to the Origin in the header gets around the issue.

**Pull request merged**